File: Cpp/LeetCode42.TrappingRainWater.cpp
// time complexity: O(N)
// space complexity: O(N)

#include <vector>

class Solution {
public:
    int trap(std::vector<int>& height) {
        int N = height.size();
        std::vector<int> arrayH(N, 0);
        if (N==0) return 0;
        // left max
        int maxH = 0;
        for (int i=1; i<N; i++) {
            maxH = std::max(maxH, height[i-1]);
            arrayH[i] = maxH;
            // Definition of arrayH[i] at here:
            // amount of water can be trapped at index j, without considering right boundary
        }
        arrayH[N-1] = 0;
        maxH = 0;
        for (int j=N-2;j>=0;j--) {
            maxH = std::max(maxH, height[j+1]);
            arrayH[j] = std::min(arrayH[j],maxH); 
            // Definition of arrayH[i] at here:
            // amount of water can be trapped at index j, after considering right boundary
            
            if (arrayH[j]>height[j]) {
                arrayH[j] -= height[j];
            } else {
                arrayH[j] = 0; // if pillar's height is greater and equal, no water can be trapped
            }
        }
        int totalArea = 0;
        for (int i=0;i<N;i++) {
            totalArea += arrayH[i];
        }
      return totalArea;
    }
};

File: Python/LeetCode42.TrappingRainWater.py
class Solution:
    def trap(self, height: List[int]) -> int:
        N = len(height)
        leftMax = height[0]
        rightMax = height[N-1]
        left = 0
        right = N-1
        total = 0

        while left < right:
            leftMax = max(leftMax, height[left])
            rightMax = max(rightMax, height[right])
            if leftMax < rightMax:
                total += leftMax - height[left]
                left += 1
            else:
                total += rightMax - height[right]
                right -= 1
        return total
